source/Missions/Services/Agents/Handlers/AgentDamageHandler.cs

```diff
@@ -15,128 +16,148 @@
+{
+
+}
+/// <inheritdoc/>
+public class AgentDamageHandler : IAgentDamageHandler
+{
+    private readonly static ILogger Logger = LogManager.GetLogger<AgentDamageHandler>();
+
+    private readonly INetworkAgentRegistry networkAgentRegistry;
+    private readonly INetworkMessageBroker networkMessageBroker;
+    private readonly INetworkMissileRegistry networkMissileRegistry;
+
+    public AgentDamageHandler(
+        INetworkAgentRegistry networkAgentRegistry,
+        INetworkMessageBroker networkMessageBroker,
+        INetworkMissileRegistry networkMissileRegistry) 
     {
+        this.networkAgentRegistry = networkAgentRegistry;
+        this.networkMessageBroker = networkMessageBroker;
+        this.networkMissileRegistry = networkMissileRegistry;
 
+        networkMessageBroker.Subscribe<AgentDamaged>(AgentDamageCheckSend);
+        networkMessageBroker.Subscribe<NetworkAgentDamaged>(AgentDamageCheck);
+        networkMessageBroker.Subscribe<ConfirmedNetworkAgentDamaged>(AgentDamageRecieve);
     }
-    /// <inheritdoc/>
-    public class AgentDamageHandler : IAgentDamageHandler
+    ~AgentDamageHandler()
     {
-        private readonly static ILogger Logger = LogManager.GetLogger<AgentDamageHandler>();
+        Dispose();
+    }
+    public void Dispose()
+    {
+        networkMessageBroker.Unsubscribe<AgentDamaged>(AgentDamageCheckSend);
+        networkMessageBroker.Unsubscribe<NetworkAgentDamaged>(AgentDamageCheck);
+        networkMessageBroker.Unsubscribe<ConfirmedNetworkAgentDamaged>(AgentDamageRecieve);
+    }
 
-        private readonly INetworkAgentRegistry networkAgentRegistry;
-        private readonly INetworkMessageBroker networkMessageBroker;
-        private readonly INetworkMissileRegistry networkMissileRegistry;
+    private void AgentDamageCheckSend(MessagePayload<AgentDamaged> payload)
+    {
 
-        public AgentDamageHandler(
-            INetworkAgentRegistry networkAgentRegistry,
-            INetworkMessageBroker networkMessageBroker,
-            INetworkMissileRegistry networkMissileRegistry) 
+        // first, check if the attacker exists in the agent to ID groud, if not, no networking is needed (not a network agent)
+        if (networkAgentRegistry.TryGetAgentId(payload.What.AttackerAgent, out Guid attackerId) == false) return;
+
+        // next, check if the attacker is one of ours, if not, no networking is needed (not our agent dealing damage)
+        if (networkAgentRegistry.IsControlled(attackerId) == false) return;
+
+        if (networkAgentRegistry.TryGetAgentId(payload.What.VictimAgent, out Guid victimId) == false)
         {
-            this.networkAgentRegistry = networkAgentRegistry;
-            this.networkMessageBroker = networkMessageBroker;
-            this.networkMissileRegistry = networkMissileRegistry;
+            Logger.Warning("Unable to get id for {agent} in {class}", payload.What.VictimAgent, typeof(AgentDamageHandler));
+            return;
+        };
+
```

Comment:
Currently friendly fire is not handled, this should fix that

```suggestion
        // Handles friend fire event
        if (networkAgentRegistry.IsControlled(victimId))
        {
            ConfirmedNetworkAgentDamaged friendlyFireMessage = new ConfirmedNetworkAgentDamaged(
                attackerId,
                victimId,
                payload.What.AttackCollisionData,
                payload.What.Blow);

            networkMessageBroker.PublishNetworkEvent(friendlyFireMessage);
            return;
        }
```